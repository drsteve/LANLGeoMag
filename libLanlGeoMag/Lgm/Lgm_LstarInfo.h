#ifndef LGM_LSTAR_INFO_H
#define LGM_LSTAR_INFO_H

#include <math.h>
#include "Lgm_QuadPack.h"
#include "Lgm_MagModelInfo.h"
#include <gsl/gsl_errno.h>
#include <gsl/gsl_spline.h>

#define ELECTRON_MASS   (9.10938188e-31)    /* Electron Mass , kg */
#define AMU             (1.660538e-27)      /* kg */
#define PROTON_MASS     (1.00794*AMU)       /* kg */
#define OXYGEN_MASS     (15.9994*AMU)       /* kg */
#define RE              (6378.135e3)        /* Earth Radius, m */
#define CC              (2.99792458e8)      /* Speed of Light, m/s */

#define LGM_DRIFT_ORBIT_CLOSED                1
#define LGM_DRIFT_ORBIT_CLOSED_SHABANSKY      2
#define LGM_DRIFT_ORBIT_CLOSED_SHABANSKY_I    21
#define LGM_DRIFT_ORBIT_CLOSED_SHABANSKY_II   22
#define LGM_DRIFT_ORBIT_CLOSED_SHABANSKY_III  23
#define LGM_DRIFT_ORBIT_OPEN                  3
#define LGM_DRIFT_ORBIT_OPEN_SHABANSKY        4
#define LGM_DRIFT_ORBIT_OPEN_SHABANSKY_I      41
#define LGM_DRIFT_ORBIT_OPEN_SHABANSKY_II     42
#define LGM_DRIFT_ORBIT_OPEN_SHABANSKY_III    43

#define LGM_SHABANSKY_IGNORE        0
#define LGM_SHABANSKY_HALVE_I       1
#define LGM_SHABANSKY_REJECT        2

#define LGM_LSTAR_MOMENT_CDIP       0
#define LGM_LSTAR_MOMENT_CDIP_2010  1
#define LGM_LSTAR_MOMENT_MCILWAIN   2

#define LGM_LSTARINFO_MAX_FL        300
#define LGM_LSTARINFO_MAX_MINIMA    300


typedef struct Lgm_LstarInfo {

    int         nFLsInDriftShell;   //!< Number of Field Lines to use when constructing Drift Shell.
    int         LstarQuality;       //!< Quality factor to use [0,8] -- higher gives more precise results.
    int         ShabanskyHandling;  //!< Set how Lstar calculations are to handle Shabansky orbits
    int         LstarMoment;        //!< Select dipole magnetic moment to use in conversion of Phi to Lstar

    double      KineticEnergy;      //!< Particle kinetic energy (only for energy dep. quantities.)
    double      Mass;               //!< Particle mass
    double      PitchAngle;         //!< Particle Pitch Angle
    double      LSimpleMax;         //!< Threshold for doing drift-shell calculation.


    Lgm_MagModelInfo	*mInfo;

    /*
     *  Variables to hold info on field lines defining the Drift Shell
     */
    int         FindShellPmin;      //!< Find the Bmin location on each FL
    int         ComputeVgc;         //!< Compute the gradient of I and Vgc
    int         SaveShellLines;     //!< only save them if this is true
    int         nFieldPnts[ LGM_LSTARINFO_MAX_FL ];    //!< number of points in each FL.
    double      s_gsm[ LGM_LSTARINFO_MAX_FL ][1000];   //!< distance along FL.
    double      Bmag[ LGM_LSTARINFO_MAX_FL ][1000];    //!< Field magnitude
    double      x_gsm[ LGM_LSTARINFO_MAX_FL ][1000];
    double      y_gsm[ LGM_LSTARINFO_MAX_FL ][1000];
    double      z_gsm[ LGM_LSTARINFO_MAX_FL ][1000];


    /*
     *  Variables to hold info on footprint of Drift Shell
     */
    int                 nPnts;
    double              MLT[ LGM_LSTARINFO_MAX_FL ], mlat[ LGM_LSTARINFO_MAX_FL ];

    Lgm_Vector          Spherical_Footprint_Pn[ LGM_LSTARINFO_MAX_FL ];
    double              Spherical_Footprint_Sn[ LGM_LSTARINFO_MAX_FL ];
    double              Spherical_Footprint_Bn[ LGM_LSTARINFO_MAX_FL ];

    Lgm_Vector          Spherical_Footprint_Ps[ LGM_LSTARINFO_MAX_FL ];
    double              Spherical_Footprint_Ss[ LGM_LSTARINFO_MAX_FL ];
    double              Spherical_Footprint_Bs[ LGM_LSTARINFO_MAX_FL ];

    Lgm_Vector          Ellipsoid_Footprint_Pn[ LGM_LSTARINFO_MAX_FL ];
    double              Ellipsoid_Footprint_Sn[ LGM_LSTARINFO_MAX_FL ];
    double              Ellipsoid_Footprint_Bn[ LGM_LSTARINFO_MAX_FL ];

    Lgm_Vector          Ellipsoid_Footprint_Ps[ LGM_LSTARINFO_MAX_FL ];
    double              Ellipsoid_Footprint_Ss[ LGM_LSTARINFO_MAX_FL ];
    double              Ellipsoid_Footprint_Bs[ LGM_LSTARINFO_MAX_FL ];

    Lgm_Vector          Mirror_Pn[ LGM_LSTARINFO_MAX_FL ];
    double              Mirror_Sn[ LGM_LSTARINFO_MAX_FL ];

    Lgm_Vector          Mirror_Ps[ LGM_LSTARINFO_MAX_FL ];
    double              Mirror_Ss[ LGM_LSTARINFO_MAX_FL ];

    double              PhiVal[ LGM_LSTARINFO_MAX_FL ], AngularVelocity[ LGM_LSTARINFO_MAX_FL ];

    double              Sb0;        // Equatorial value of Sb Integral.
    double              d2B_ds2;    // second derivative of B wrt s at equator.
    double              RofC;       // radius of curvature at Bmin point.

    double              I0;
    double              I[ LGM_LSTARINFO_MAX_FL ];

    int                 ComputeSbIntegral;
    double              SbIntegral0; // Sb Integral on initial FL (i.e. one with Sat through it.)
    //double              SbIntegral[ LGM_LSTARINFO_MAX_FL ]; // havent added code to compute these yet (do we need to?)

    Lgm_Vector          Bmin[ LGM_LSTARINFO_MAX_FL ];
    Lgm_Vector          Pmin[ LGM_LSTARINFO_MAX_FL ];
    Lgm_Vector          GradI[ LGM_LSTARINFO_MAX_FL ];
    Lgm_Vector          Vgc[ LGM_LSTARINFO_MAX_FL ];

    int                 DriftOrbitType;         // e.g. Open, Closed, Shabansky
    int                 nMinMax;                // Number of valid FLs represented in nMinima[] and nMaxima[] (we may have bailed early)
    int                 nMinima[ LGM_LSTARINFO_MAX_MINIMA ];           // # of minima on FL
    int                 nMaxima[ LGM_LSTARINFO_MAX_MINIMA ];           // # of maxima on FL (not including endpoints)
    int                 nBounceRegions[ LGM_LSTARINFO_MAX_MINIMA ];    // # of bounce regions on FL (as particle may not be confined in a minimum)

    int                 nSplnPnts;
    double              xa [ 3*LGM_LSTARINFO_MAX_FL ], ya[ 3*LGM_LSTARINFO_MAX_FL ], y2[ 3*LGM_LSTARINFO_MAX_FL ];

    double	        Phi;
    double              Mused;   // Magnetic moment to use in Lstar

    gsl_interp_accel    *acc;
    gsl_interp          *pspline;



    /*
     *  Other stuff
     */
    int		            VerbosityLevel;
    char                PreStr[64], PostStr[64];
    int                 ISearchMethod;

    double	            LS;
    double	            LS_dip_approx;
    double	            LS_McIlwain_M;

    int                 m;
    double              xma[ 3*LGM_LSTARINFO_MAX_FL ], yma[ 3*LGM_LSTARINFO_MAX_FL ], ym2[ 3*LGM_LSTARINFO_MAX_FL ];


    /*
     * Variables to keep track of the I-I0 versus mlat variations that result
     * in search attempts in FindShellLine(). These vars are used in Lstar() to
     * help figure out ranges of mlat to search over.
     */
    double  MLATarr[ 3*LGM_LSTARINFO_MAX_FL ]; // mlat's
    double  ImI0arr[ 3*LGM_LSTARINFO_MAX_FL ]; // I-I0's
    double  Earr[ 3*LGM_LSTARINFO_MAX_FL ];    // nominally the error on I-I0 (typically set to const).
    int     nImI0;        // number of vals stored.



    /*
     *  variables for keeping track of particles
     */
    int                 nParticles;
    Lgm_Vector          Particles[5000];


} Lgm_LstarInfo;


void        Lgm_SetLstarTolerances( int Quality, int nFLsInDriftShell, Lgm_LstarInfo *LstarInfo );
Lgm_LstarInfo  *InitLstarInfo( int VerbosityLevel );
//void Lgm_InitMagInfoDefaults( Lgm_MagModelInfo  * );
void Lgm_InitLstarInfoDefaults( Lgm_LstarInfo   *LstarInfo );

void FreeLstarInfo( Lgm_LstarInfo *LstarInfo );
Lgm_LstarInfo *Lgm_CopyLstarInfo( Lgm_LstarInfo *s );

int         Grad_I( Lgm_Vector *vin, Lgm_Vector *GradI, Lgm_LstarInfo *LstarInfo );
int         ComputeVcg( Lgm_Vector *vin, Lgm_Vector *Vcg, Lgm_LstarInfo *LstarInfo );
int 	    FindBmRadius( double Bm, double MLT, double mlat, double *r, double tol, Lgm_LstarInfo *LstarInfo );
int 	    FindShellLine( double I0, double *Ifound, double Bm, double MLT, double *mlat, double *rad, double mlat0, double mlat1, double mlat2, int *Iterations, Lgm_LstarInfo *LstarInfo );
double      ComputeI_FromMltMlat(  double Bm, double MLT, double mlat, double *r, double I0, Lgm_LstarInfo *LstarInfo );
double      ComputeI_FromMltMlat1( double Bm, double MLT, double mlat, double *r, double I0, Lgm_LstarInfo *LstarInfo );
double      ComputeI_FromMltMlat2( double Bm, double MLT, double mlat, double *r, double I0, Lgm_LstarInfo *LstarInfo );
void 	    spline( double *x, double *y, int n, double yp1, double ypn, double *y2);
void 	    splint( double *xa, double *ya, double *y2a, int n, double x, double *y);
void 	    quicksort( unsigned long n, double *arr );
void 	    quicksort2(  unsigned long n, double *arr, double *brr );
void 	    quicksort3(  unsigned long n, double *arr, double *brr, double *crr );
Lgm_MagModelInfo     *init_info( ) ;
void        NewTimeLstarInfo( long int Date, double UT, double PitchAngle, int (*Mag)(Lgm_Vector *, Lgm_Vector *, Lgm_MagModelInfo *), Lgm_LstarInfo *LstarInfo );
int         Lstar( Lgm_Vector *vin, Lgm_LstarInfo *LstarInfo );
double      MagFlux( Lgm_LstarInfo *LstarInfo );
double      MagFlux2( Lgm_LstarInfo *LstarInfo );
double      MagFluxIntegrand( double Phi, _qpInfo *qpInfo ) ;
double      MagFluxIntegrand2( double Phi, _qpInfo *qpInfo ) ;
double      LambdaIntegrand( double Lambda, _qpInfo *qpInfo ) ;
double      LambdaIntegral( Lgm_LstarInfo *LstarInfo ) ;
double      AngVelInv( double Phi );
int         Lgm_LCDS( long int Date, double UTC, double brac1, double brac2, double Alpha, double LT, double tol, int Quality, int nFLsInDriftShell, double *K, Lgm_LstarInfo *LstarInfo );
 

#endif
