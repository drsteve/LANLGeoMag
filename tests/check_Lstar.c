#include <check.h>
#include "../libLanlGeoMag/Lgm/Lgm_LstarInfo.h"
#include "../libLanlGeoMag/Lgm/Lgm_MagEphemInfo.h"


Lgm_LstarInfo *LstarInfo;


void Lstar_Setup(void) {
    LstarInfo = InitLstarInfo(1);
    return;
}

void Lstar_TearDown(void) {
    FreeLstarInfo(LstarInfo);
    return;
}

START_TEST(test_Lstar_CDIP){
    /*
     *  Lstar in centered dipole should give known result
     */

    double           UTC, Blocal, sa, sa2, LstarDiff, tol;
    long int         Date;
    int              Kp, quality;
    Lgm_Vector       Psm, P, Bvec, v1, v2, v3;

    // Date and UTC
    Date       = 19891020;
    UTC        = 21.0;
    Lgm_Set_Coord_Transforms( Date, UTC, LstarInfo->mInfo->c );

    // Position in SM
    Psm.x = 0.0; Psm.z = 0.0;
    Psm.y = 7.0;
    Lgm_Convert_Coords( &Psm, &P, SM_TO_GSM, LstarInfo->mInfo->c );

    // pitch angles to compute
    LstarInfo->PitchAngle = 90.0;
    LstarInfo->LstarMoment = LGM_LSTAR_MOMENT_CDIP;
    Lgm_MagModelInfo_Set_MagModel( LGM_CDIP, LGM_EXTMODEL_NULL, LstarInfo->mInfo );
    quality = 3;
    Lgm_SetLstarTolerances( quality, 48, LstarInfo );

    Lstar( &P, LstarInfo );

    LstarDiff = fabs(LstarInfo->LS - Psm.y);

    tol = pow(10.0, (double) -quality );
    ck_assert((LstarDiff<tol));

    return;

}END_TEST


START_TEST(test_Lstar_CDIPapprox){
    /*
     *  Lstar in centered dipole should give same result with full field and dipole approx
     */

    double           UTC, Blocal, sa, sa2, LstarDiff, tol;
    long int         Date;
    int              Kp, quality;
    Lgm_Vector       Psm, P, Bvec, v1, v2, v3;

    // Date and UTC
    Date       = 20110123;
    UTC        = 3.14;
    Lgm_Set_Coord_Transforms( Date, UTC, LstarInfo->mInfo->c );

    // Position in SM
    Psm.x = 0.0; Psm.z = 0.0;
    Psm.y = 6.0;
    Lgm_Convert_Coords( &Psm, &P, SM_TO_GSM, LstarInfo->mInfo->c );

    // pitch angles to compute
    LstarInfo->PitchAngle = 75.0;
    LstarInfo->LstarMoment = LGM_LSTAR_MOMENT_CDIP;
    Lgm_MagModelInfo_Set_MagModel( LGM_CDIP, LGM_EXTMODEL_NULL, LstarInfo->mInfo );
    quality = 3;
    Lgm_SetLstarTolerances( quality, 36, LstarInfo );

    Lstar( &P, LstarInfo );

    LstarDiff = fabs(LstarInfo->LS - LstarInfo->LS_dip_approx);

    tol = pow(20.0, (double) -quality );
    ck_assert_msg((LstarDiff<tol), "LS = %g; LS_dip_approx = %g; LstarDiff = %g; Mused = %g; Mcd = %g; Mcd2010 = %g\n", LstarInfo->LS, LstarInfo->LS_dip_approx, LstarDiff, LstarInfo->Mused, LstarInfo->mInfo->c->M_cd, LstarInfo->mInfo->c->M_cd_2010);

    return;

}END_TEST


START_TEST(test_Lstar_CDIPalpha){
    /*
     *  Same pitch angle in centered dipole should give same result
     */

    double           UTC, Blocal, sa, sa2, LstarDiff, ans1, ans2, tol;
    long int         Date;
    int              Kp, quality;
    Lgm_Vector       Psm, P, Bvec, v1, v2, v3;

    // Date and UTC
    Date       = 19991122;
    UTC        = 19.0;
    Lgm_Set_Coord_Transforms( Date, UTC, LstarInfo->mInfo->c );

    // Position in SM
    Psm.x = 0.0; Psm.y = 6.6; Psm.z = 0.0;
    Lgm_Convert_Coords( &Psm, &P, SM_TO_GSM, LstarInfo->mInfo->c );

    // set up Lstar calcs
    LstarInfo->LstarMoment = LGM_LSTAR_MOMENT_CDIP_2010;
    Lgm_MagModelInfo_Set_MagModel( LGM_CDIP, LGM_EXTMODEL_NULL, LstarInfo->mInfo );
    Lgm_SetLstarTolerances( 5, 24, LstarInfo );

    // run 1
    LstarInfo->PitchAngle = 87.5;
    Lstar( &P, LstarInfo );
    ans1 = LstarInfo->LS;

    // run 2
    LstarInfo->PitchAngle = 37.5;
    Lstar( &P, LstarInfo );
    ans2 = LstarInfo->LS;

    // compare
    LstarDiff = fabs(ans2-ans1);
    tol = pow(20.0, (double) -quality );

    ck_assert_msg((LstarDiff<tol), "Roederer L differs in CDIP for different pitch angles. %g - %g = %g\n", ans2, ans1, LstarDiff);


    return;

}END_TEST


START_TEST(test_Lstar_CDIPalpha2){
    /*
     *  Should get same result from Lstar and ComputeLstarVersusPA
     */

    double           UTC, Blocal, sa, sa2, LstarDiff, ans2, tol;
    double           PAs[2] = {87.5, 37.5};
    long int         Date;
    int              Kp, quality=3;
    Lgm_Vector       Psm, P, Bvec;
    Lgm_MagEphemInfo *MagEphemInfo = Lgm_InitMagEphemInfo(1, 2);

    // Date, UTC, position
    Date       = 19991122;
    UTC        = 19.0;
    Psm.x = 0.0; Psm.y = 6.6; Psm.z = 0.0;

    /*
     * Calculate using Lstar directly
     */
    Lgm_Set_Coord_Transforms( Date, UTC, LstarInfo->mInfo->c );

    // Position in SM
    Lgm_Convert_Coords( &Psm, &P, SM_TO_GSM, LstarInfo->mInfo->c );

    // set up Lstar calcs
    LstarInfo->LstarMoment = LGM_LSTAR_MOMENT_CDIP_2010;
    Lgm_MagModelInfo_Set_MagModel( LGM_EDIP, LGM_EXTMODEL_NULL, LstarInfo->mInfo );
    Lgm_SetLstarTolerances( quality, 72, LstarInfo );

    // Get Lstar
    LstarInfo->PitchAngle = PAs[1];
    Lstar( &P, LstarInfo );
    ans2 = LstarInfo->LS;

    /*
     * Calculate using Lgm_ComputeLstarVersusPA
     */

    Lgm_SetMagEphemLstarQuality( quality, 72, MagEphemInfo );
    MagEphemInfo->LstarInfo->LstarMoment = LGM_LSTAR_MOMENT_CDIP_2010;
    Lgm_MagModelInfo_Set_MagModel( LGM_EDIP, LGM_EXTMODEL_NULL, MagEphemInfo->LstarInfo->mInfo );

    Lgm_ComputeLstarVersusPA( Date, UTC, &P, 2, PAs, FALSE, MagEphemInfo );

    printf("Lstar (from Lstar) = %g\nLstar (from ComputeLstarVersusPA) = %g\n", ans2, MagEphemInfo->Lstar[1]);

    // compare
    LstarDiff = fabs(ans2 - MagEphemInfo->Lstar[1]);
    tol = pow(20.0, (double) -quality );

    ck_assert_msg((LstarDiff<tol), "Roederer L differs between Lstar and ComputeLstarVersusPA. %g - %g = %g\n", ans2, MagEphemInfo->Lstar[1], LstarDiff);


    return;

}END_TEST


START_TEST(test_Lstar_Regressions) {
    /* Regression tests against previous L* results */
    Lgm_Vector        Pos, PosGSM;
    int               nTests, nPass, nFail, transflag, SubtestPassed, \
                      Passed=FALSE;
    double            del, Kp, PA, HiltonExpect, HiltonTest, McIlwainExpect, \
                      McIlwainTest, RoedererExpect, RoedererTest, I, Bm, M;
    char              buff[262], extModel[10], intModel[10];
    char              IsoDate[80];
    FILE              *testfile, *outfile;
    Lgm_DateTime      d;

    int makeNew = 1;

    /* read test file */
    testfile = fopen("check_Lstar.expected","r");
    if (makeNew) outfile = fopen("check_Lstar.got", "w");

    /* step through test cases one line at a time */
    nTests = 0;
    nPass = 0;
    nFail = 0;
    Passed = TRUE;
    while( fgets(buff,260,testfile) != NULL) {
        if (buff[0]!='#') {
	    SubtestPassed = TRUE;
            // read line
            sscanf(buff,
		   "%s %s %s %lf "
		   "%lf %lf %lf "
		   "%lf %lf %lf %lf",
		   IsoDate, extModel, intModel, &Kp, 
                   &Pos.x, &Pos.y, &Pos.z,
		   &PA, &HiltonExpect, &McIlwainExpect, &RoedererExpect);
            IsoTimeStringToDateTime( IsoDate, &d, LstarInfo->mInfo->c );
            //printf("IsoDate = %s; d.Date, d.Time = %ld, %lf \n", IsoDate, d.Date, d.Time);
            Lgm_Set_Coord_Transforms( d.Date, d.Time, LstarInfo->mInfo->c );
	    //Assume input is SM (since it is in Python tests)
	    Lgm_Convert_Coords(&Pos, &PosGSM, SM_TO_GSM, LstarInfo->mInfo->c );
            nTests++;
	    if (!strncmp(intModel, "IGRF", 10)) {
	        Lgm_Set_Lgm_B_IGRF_InternalModel( LstarInfo->mInfo );
	    }
	    else if (!strncmp(intModel, "CDIP", 10)) {
	        Lgm_Set_Lgm_B_cdip_InternalModel( LstarInfo->mInfo );
	    }
	    else if (!strncmp(intModel, "EDIP", 10)) {
	        Lgm_Set_Lgm_B_edip_InternalModel( LstarInfo->mInfo );
	    }
	    else {
	        nFail++;
	        printf("Test %d bad internal model %s\n", nTests, intModel);
		continue;
	    }
	    if (!strncmp(extModel, "T89", 10)) {
	        Lgm_Set_Lgm_B_T89(LstarInfo->mInfo);
	    }
	    else if (!strncmp(extModel, "OP77", 10)) {
	        Lgm_Set_Lgm_B_OP77(LstarInfo->mInfo);
	    }
	    else if (!strncmp(extModel, "NULL", 10)) {
	        LstarInfo->mInfo->ExternalModel = LGM_EXTMODEL_NULL;
	    }
	    else {
	        nFail++;
	        printf("Test %d bad external model %s\n", nTests, extModel);
		continue;
	    }
	    LstarInfo->mInfo->Kp = Kp;
	    LstarInfo->PitchAngle = PA;
	    Lgm_SetLstarTolerances( 1, 24, LstarInfo );
	    //	    LstarInfo->ShabanskyHandling = LGM_SHABANSKY_IGNORE;
	    McIlwainTest = Lgm_McIlwain_L(d.Date, d.Time, &PosGSM,
					  PA, 0, &I, &Bm, &M, LstarInfo->mInfo);
	    HiltonTest = Lgm_McIlwain_L(d.Date, d.Time, &PosGSM,
					PA, 1, &I, &Bm, &M, LstarInfo->mInfo);
	    Lstar(&PosGSM, LstarInfo);
	    RoedererTest = LstarInfo->LS;
	    del = HiltonTest - HiltonExpect;
	    if (fabs(del) > 1.0e-5) {
	        SubtestPassed = FALSE;
                printf("*****  warning : Hilton difference >= 1.0e-5 *****\n");
                printf("Test %d failed (diff: %g)\n", nTests, del);
            }
	    del = McIlwainTest - McIlwainExpect;
	    if (fabs(del) > 1.0e-5) {
	        SubtestPassed = FALSE;
                printf("*****  warning : McIlwain difference >= 1.0e-5 *****\n");
                printf("Test %d failed (diff: %g)\n", nTests, del);
            }
	    del = RoedererTest - RoedererExpect;
	    if (fabs(del) > 1.0e-5) {
	        SubtestPassed = FALSE;
                printf("*****  warning : Roederer difference >= 1.0e-5 *****\n");
                printf("Test %d failed (diff: %g)\n", nTests, del);
            }
	    if (SubtestPassed) {
	        nPass++;
                printf("Test %d passed\n", nTests);
            }
	    else {
	        nFail++;
	    }
            if (makeNew) fprintf(
		   outfile,
		   "%s %s %s %lf "
		   "%lf %lf %lf "
		   "%lf %lf %lf %lf\n",
		   IsoDate, extModel, intModel, Kp,
		   Pos.x, Pos.y, Pos.z,
		   PA, HiltonTest, McIlwainTest, RoedererTest);
            }
        else {
            if (makeNew) fprintf(outfile, "%s", buff);
            }
        }
    if (nFail>0) Passed = FALSE;
    fclose(testfile);
    if (makeNew) fclose(outfile);
    printf("Result: %d tests pass; %d tests fail (Precision=1.0e-5)\n", nPass, nFail);
    fflush(stdout);

    ck_assert_msg( Passed, "LstarRegressions tests failed.\n" );

}
END_TEST


Suite *Lstar_suite(void) {

  Suite *s = suite_create("ROEDERER_L_TESTS");

  TCase *tc_Lstar = tcase_create("Roederer L tests");
  tcase_add_checked_fixture(tc_Lstar, Lstar_Setup, Lstar_TearDown);

  tcase_add_test(tc_Lstar, test_Lstar_CDIP);
  tcase_add_test(tc_Lstar, test_Lstar_CDIPapprox);
  tcase_add_test(tc_Lstar, test_Lstar_CDIPalpha);
  tcase_add_test(tc_Lstar, test_Lstar_CDIPalpha2);
  tcase_add_test(tc_Lstar, test_Lstar_Regressions);

  suite_add_tcase(s, tc_Lstar);

  return s;

}

int main(void) {

    int      number_failed;
    Suite   *s  = Lstar_suite();
    SRunner *sr = srunner_create(s);

    printf("\n\n");
    srunner_run_all(sr, CK_NORMAL);
    number_failed = srunner_ntests_failed(sr);
    srunner_free(sr);

    return (number_failed == 0) ? EXIT_SUCCESS : EXIT_FAILURE;

}
