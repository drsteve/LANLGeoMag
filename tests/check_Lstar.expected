#{
#    "CreationDate": "2019-04-12",
#    "LanlGeoMagRevision": "122e2a8b179210d7d5d7e7d332bb03fa24bc9a10",
#    "SunVector": "LOW_ACCURACY",
#    "PA": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 7
#    },
#    "DateTime": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 0
#    }, 
#    "ExtModel": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 1
#    }, 
#    "Hilton": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 8
#    },
#    "IntModel": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 2
#    },
#    "Kp": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 3
#    },
#    "McIlwain": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 9
#    },
#    "Roederer": {
#        "DIMENSION": [1], 
#        "START_COLUMN": 10
#    },
#    "TestPosition": {
#        "DIMENSION": [3], 
#        "START_COLUMN": 4, 
#        "units": "km"
#    }
#}
2010-10-12T00:00:00.000000 T89 IGRF 4. -4.2 1.0 1.0 90. 5.029270900000001 5.029392209300000 4.584505259800000
2010-10-12T00:00:00.000000 T89 IGRF 5. -4.2 1.0 1.0 90. 5.181063798000000 5.181235930000001 4.600643183400000
2010-10-12T00:00:00.000000 OP77 IGRF 4. -4.2 1.0 1.0 90. 4.864068136599998 4.864149869999999 4.607683093099999
2010-10-12T00:00:00.000000 OP77 IGRF 5. -4.2 1.0 1.0 90. 4.862572936599999 4.862654030000000 4.607683093099999
2010-10-12T00:00:00.000000 NULL CDIP 2. -3.0 0.0 0.0 90. 3.04107 3.04107 2.981669622000000
